### ISSUE TITLE
convergence fault in the presence of non-segwit signaling blocks

### ISSUE THREAD

**Issue by gmaxwell** [btc1/bitcoin#85]
Alice, Bob, Carol, Dave, etc. are 'btc1' miners and have begun enforcing segwit (bit1) signaling.

These miners are not connected directly to each other but only via the network of ordinary Bitcoin nodes, as is likely.

Mallory is an "unlimited" running miner (or at least pretending to be) and not signaling segwit.

Mallory creates a non-BIT1 signaling block.

A, B, C, D... all ignore her block because it doesn't signal.

Alice mines a block. But B, C, D, etc. won't hear it because it isn't the longest chain to the rest of the network.

Bob mines a block but A, C, D, etc. won't hear it... and so on.

Meanwhile Mallory and the other non-btc1 miners keep on mining more blocks, all collaborating on a single chain.

So the network breaks into as many forks as there are partitioned 'btc1' miners until a _single_ btc1 miner overtakes all the non-btc1 miners combined, which may not even be possible.

Resolution is to either guarantee there are non-bit1 signaling miners before enforcing (kind of incompatible with the definition of 80% activation), or manually guarantee the the graph of enforcing miners is connected.

This requires an urgent response.

**Comment by jameshilliard** [comment]
Miners should be ok as long as they are connected to the relay network.

Per @TheBlueMatt:

```
Hey y'all,

Please keep in mind that neither James' segsignal nor Jeff's btc1 have
any logic to ensure they connect to other BIP 91 nodes. Because such
nodes are very few, and because other nodes will *not* relay blocks
which are not their best chain, it is very possible that your nodes will
not receive notifications of new blocks on the BIP 91 chain.

This means that any time a miner mines a block which does meet the BIP
91 requirements, the fork block which does will take a longer amount of
time to reach your nodes, if you see it at all!

Please, please make absolutely sure that your BIP 91-enforcing nodes
have connections (via -addnode=$NODE or addnode $NODE add) too other BIP
91 miners.

Additionally, the public FIBRE nodes will be adapted later today to
relay all blocks which they receive which are recent. This should be
sufficient to solve the issue as well.

In conclusion, MAKE SURE you are connected to the public FIBRE network
as well as other miners BEFORE the BIP 91 activation takes place. Keep
in mind that bitcoind is limited to 8 addnodes (this limit is separate
from the outbound connections limit).

Thanks and good luck,
Matt
```

**Comment by TheBlueMatt** [comment]
Well there's gonna be some changes needed to the relay network to make it work, and ideally you wouldnt trust my rather ad-hoc network to stay online, but, yes, hopefully. Still, it is strongly, strongly recommended that non-miners do *not* run btc1/bip 91 until well after the bip 91 fork.

**Comment by JaredR26** [comment]
This seems like an issue with BIP 148/91's approach, correct?  The only
reason signaling rejection was added in was by request for compatibility
with those, bit4 required signalling was never merged(I think).

> as there are partitioned 'btc1' miners
> until a *single* btc1 miner overtakes all
> the non-btc1 miners combined

In reality most of the big pools are direct peers and/or the fast backbone
network.


That said, this does appear to be a legitimate concern, thank you for
raising it Greg.


It seems that manually ensuring direct peering is the best resolution of
this?


Is there a code approach that should have been taken instead (for future
documentation/proposals if nothing else)?

Jared

**Comment by TheBlueMatt** [comment]
This isnt really an issue with BIP148, cause they have lots of nodes and good connectivity, this is, however, a major issue for btc1.

> In reality most of the big pools are direct peers and/or the fast backbone network.

Not really true anymore, sadly. This absolutely needs to change *prior* to activation.

**Comment by jameshilliard** [comment]
> This seems like an issue with BIP 148/91's approach, correct?

It's more of an issue just because there aren't many BIP91 nodes, but it should be manageable as long as pool operators make sure they are connected properly to other pools.

**Comment by jgarzik** [comment]
Agree it's an issue.  The new DNS seeds do inject a heavier dose of 2x nodes; whether or not that has a positive impact depends on your situation (a weak help, IOW).

**Comment by TheBlueMatt** [comment]
@jameshilliard yes, this is why it is very strongly recommended that *only* miners (who should all have careful setups) run bip91/btc1 until well after the bip 91 fork!

@jgarzik there is code which makes sure that dnsseeds are not queried almost at all (except very first startup), so, no, that isnt just a weak help, its no help.

**Comment by JaredR26** [comment]
Taking that statement to be true, another solution would be to add many
more btc1 nodes, correct?

(In parallel with the manual peering changes)

**Comment by TheBlueMatt** [comment]
Not really, because the btc1 nodes dont contain the 14.2 fixes, so peer-finding is further broken (seriously, why?) and because you need time to propagate information about those nodes through the network.

**Comment by gmaxwell** [comment]
> The new DNS seeds do inject a heavier dose of 2x nodes; 

What seeds add to your pool doesn't really do anything there because the ordinary seeds aren't filtering them out.  They'll be no more likely to get selected than any other node.  A quick test with 5 restarts shows every try resulted in node fully partitioned.

BIP 148 doesn't appear to have this issue, and would actually heal this problem if the activation were concurrent. However it looks at least somewhat likely that bit4 will activate well ahead of BIP148, and if so they won't be a help.

Aside, your hardcoded seeds returning a small set of static results is a gross violation of the DNS seed policy copied over from the Bitcoin project.

**Comment by earonesty** [comment]
If 0.14.2 were merged in, it would allow the btc1 nodes to self-select a connected network, correct?

**Comment by TheBlueMatt** [comment]
@earonesty No, that would help a small amount, but its much too late to get the nodes to be well-propagated at this point.

**Comment by earonesty** [comment]
Well, maybe NYA miners should just run the UASF code itself.  Seems like a safer way of activating.   Then they can switch to the btc1 branch for the hard fork later.

**Comment by mjamin** [comment]
I'm having trouble imagining Jihan Wu with an UASF hat.

**Comment by adaseb** [comment]
Yeah I can see the 2MB Hard-fork going very smoothly!

**Comment by jgarzik** [comment]
Please keep the discussion technical and focused.

**Comment by jheathco** [comment]
Just a thought, but with regards to segwit2x BIP91 lock-in, couldn't the code be modified to time the orphaning of blocks to that of BIP148's activation rather than using block X+336?

This would at least allow segwit2x to team up with the BIP148 nodes in orphaning any blocks not compliant with BIP141.

**Comment by TheBlueMatt** [comment]
@jheathco That seems like a good idea, but it is *way* too late to be changing consensus rules post-release. This is precisely why you dont rush things like this, you end up regretting your decisions (see also P2SH, etc).

**Comment by kek-coin** [comment]
@jheathco how would that work in case BIP91 locks in after the BIP148 deadline?

**Comment by JaredR26** [comment]
On that note, is there a reason that the Bitcoin Core 14.2 changes can't be
brought forward in a post August 1st merge?  I might be able to work
towards that when I am home from vacation if there is support for it among
the segwit2x team.

**Comment by jheathco** [comment]
@kek-coin I'd imagine you could simply default back to block X+336 in the case that lock-in happens after BIP148 activates.

**Comment by jgarzik** [comment]
@JaredR26 Yep - it was on the mental todo list - filed as #88

**Comment by tomasvdw** [comment]
This is trivially solved by some two or three big miners making sure they are connecting to each other.

In fact, this is already solved by the largest SegWit2x miner having more hashpower than all non-segwit2x miners combined. 

Or as Gregory writes:

> So the network breaks into as many forks as there are partitioned 'btc1' miners until a single btc1 miner overtakes all the non-btc1 miners combined, which may not even be possible.

Of course this is possible. The opposite is hardly possible given the current hashrate stats.

**Comment by nukebloodaxe** [comment]
@tomasvdw Although technically right in a way, it is of cold comfort to the miners who find themselves building on the wrong partitioned chain.  I can't imagine the issues it'll cause for early end users/exchanges as well, because of sly double spending by those utilising each partitioned chain.

Really a list of primary nodes interconnected with each other needs to be made now, and then for each mining team to manually connect to at least one of them; I know this has been indicated earlier, but it really needs emphasis.

**Comment by marthinus-engelbrecht** [comment]
Noob question. How serious is this? Are we talking a potential accidental fork.

**Comment by JavierGonzalez** [comment]
@gmaxwell You should resign *.

This risk you are mentioning is the exact result of your poor management, rejecting the kindly mean term "segwit + 2MB" and artificially limiting the transactions per second.

If you do not know how to make Bitcoin scalable without breaking the community, you must resign immediately.

You still have the opportunity to do it well.
Please give an honorable end to this.

* The button to resign is to assign as administrator to Gavin Anddressen.

**Comment by tomasvdw** [comment]
We have to remain realistic. Currently the largest btc1 miner has 22%. I am trying to  imagine a situation where

A. No btc1 route exists between this miner and another btc1 miner.
B. No connected set of btc1 miners exists with more then 22%
C. More than 22% of the (non-segwit) miners decide _after BIP91 lock in_ to continue mining $25k blocks that they know will be orphaned.

Yes. It is a good idea to ping some miners to manually connect, but the scenario is not realistic.

**Comment by JavierGonzalez** [comment]
@gmaxwell You should resign (*).

This risk you are mentioning is the exact result of your poor management, rejecting the kindly mean term "segwit + 2MB" and artificially limiting the transactions per second.

If you do not know how to make Bitcoin scalable without breaking the community, you must resign immediately.

You still have the opportunity to do it well.
Give an honorable end to this.

The button to resign is to assign as administrator to Gavin Anddressen.

**Comment by Lejitz** [comment]
All of this discussion is centered around the idea that every miner purportedly running btc1 will actually enforce and adhere to the bit 1 mandate after activation.  But with false signaling, there could be quite a few that do not (even accidentally).  Imagine the havoc that could wreak.  Many nodes would follow the false signalers that mine blocks not signaling on bit 1, leaving the honest miners with little/no network support.  I'm not sure that even a list of nodes would help too much in this scenario.

All of these problems would be fixed (or most alleviated) if the code were slightly modified to make activation coincide with BIP148 activation if BIP91 lock-in occurs before that time.  If lock-in does not occur before that date, it could simply follow the X+336 schedule.

**Comment by jgarzik** [comment]
@JavierGonzalez btc1 is not the forum for that type of comment.  We are trying to build a more civil and professional forum, and that does not befit the welcome sign graphic at the door.

**Comment by JavierGonzalez** [comment]
@jgarzik Okay, I deleted my suggestion.

Now you can continue the conversation about something that is not under your control because you do not have enough trillions of hashes per second.

**Comment by kek-coin** [comment]
@JavierGonzalez peering issues can be solved without hashrate.

**Comment by JavierGonzalez** [comment]
@kek-coin This is not a simple "peering issues".

It is the prelude to a great and irreversible lose-to-lose catastrophe.

**Comment by jgarzik** [comment]
~~This issue needs a chill-out period :)~~  ETA, unlocked.

Agree w/ @tomasvdw that this can be - and is - being solved with good peering.

And if I may put words in @kek-coin 's mouth, btc1 peering issues should be attended to irrespective of hashrate.  We can solve related items already on the board (e.g. #86 ) along the way.

**Comment by starrymoonlight** [comment]
> BIP 148 doesn't appear to have this issue, and would actually heal this problem if the activation were concurrent. However it looks at least somewhat likely that bit4 will activate well ahead of BIP148, and if so they won't be a help.

Given the dangers of BIP91, if a significant group of miners could be convinced, wouldn't it be safer for miners to use BIP148 for activation? If so will btc1 consider merging in BIP148?

**Comment by kek-coin** [comment]
An idea I've seen floating in this thread would be to change the orphaning precondition such that if BIP91 activates before August 1st, BIP148-nodes will help out to alleviate the partitioning danger, while with a post-August 1st activation it would be on its own. This could be easily implemented by changing the check from "BIP91 ACTIVE" to "BIP91 ACTIVE && BIP148 ACTIVE".

Changing consensus rules on such short notice may be undesirable but since there has not yet been an official release and since BIP91 is (imho) primarily relevant for miners - and therefore users can simply keep running Core (or BIP148) while it is being sorted out - this could be a viable mitigation strategy.

**Comment by istrau2** [comment]
@kek-coin That sounds like a great solution.

**Comment by JaredR26** [comment]
> @kek-coin <https://github.com/kek-coin> That sounds like a great solution.

UASF is not a solution to BIP91.

**Comment by kek-coin** [comment]
@JaredR26 please re-read my suggestion.

**Comment by JaredR26** [comment]
Oh, my bad, I had a different comment in my head when I wrote that from
higher up.  Sorry

It isn't a terrible idea, but I think (echoing what someone on the email
list wrote) that trying to change course now might make the issue worse if
any miners don't update twice.  If they peer as specified the issue will be
mitigated.  If we had more time that might be a better approach.

**Comment by istrau2** [comment]
@JaredR26 
I see where you are coming from (changing the rules at this point is indeed risky), however I disagree. I think the change suggested by @kek-coin is quite minor and greatly increases the chance that the SF will go smoothly. Trying to fix this issue using some kind of peering hack is riskier in my opinion (and also significantly increases the burden on the miners in upgrading). 

Additionally,  (while I may be misunderstanding something here), It seems to me that we are forgetting that we also want small miners to upgrade to BIP91 compatible software (at least eventually). Its all great that we can hack together the large miners and avoid serious network issues, however these smaller miners are much more likely to be isolated (and therefore experience orphaning issues). Small miners will be very hesitant to upgrade to BIP91 (as well they should be) and this will lengthen the time period in which this issue exists as a real risk after the SF.

**Comment by kek-coin** [comment]
@JaredR26 to circle back on your psychological approach to bitcoin consensus rules; don't you think that miners and users joining forces to avoid a chainsplit might make for a mighty fine symbolic event?

**Comment by istrau2** [comment]
Actually, thinking about this more. It seems to me that the users are actually the biggest risk here. In other words if Tom is a user that is not directly connected to any BIP91 node, then even if there is a 80/20 split between BIP91 and non BIP91, isn't there a decent chance that Tom could possibly get 3 confirmations to a block that is on the non-BIP91 chain? 

I mean, the non-BIP91 nodes would keep building on top of the BIP91 blocks. All that is required is for the non-BIP91 nodes to build 3 blocks (with their 20% hashrate) before the BIP91 nodes can build 3 blocks AT ANY POINT IN TIME after the SF. Then, Tom, would see 3 confirmations to a transaction that would get wiped out pretty soon afterwards.

Am I thinking about this incorrectly?

**Comment by kek-coin** [comment]
Users and (small) miners not directly connected to the biggest BIP91 partition are at risk, yes.

**Comment by istrau2** [comment]
Additionally, I am not sure that preferential peering would be a good solution. It seems like that could open BIP91 up to malicious fake signalling attacks.

**Comment by christophebiocca** [comment]
>  then even if there is a 80/20 split between BIP91 and non BIP91, isn't there a decent chance that Tom could possibly get 3 confirmations to a block that is on the non-BIP91 chain?

This follows the same distribution as someone trying to do a 51% attach with minority hashpower, which is in the whitepaper.

So for 20% of hashpower there's a ~20% chance that after mining one block, the non-signalling and/or non-orphaning miners manage to extend their lead by 2 more blocks before the network catches up to them.

With the more realistic 12% of non-segwit2x miners, you're looking at a 7% chance of such a branch, per time the 12% mines a non-segwit block on top of the longest chain (which, since GBminers and ~1/2 of slush signal anyways, is down to KanoCK pool and a few other, totalling ~5%).

But while there's a risk for users, it's worth pointing out that each such block mined, and any block mined on top will be completely orphaned eventually.

That's a lost reward of $30k each time.

So the complete list of who's at risk is:

- Non-signalling miners: $30k for each block mined.
- Signalling-non-orphaning miners: $30k each time they mine a block on top of a non-signalling block, which will occur about 5% of the time.
- Non-peered signalling-and-orphaning miners: $30k for each block they mine trying to orphan an invalid chain, but that they don't get relayed to the 80% group.
- Users not enforcing BIP91: Possibility of accepting payments from a branch that will be orphaned, which will happen 5% of the time for 1-confirmation, ~1.25% of the time for 2 confirmations, and ~0.4% of the time for 3 confirmations.
- Users enforcing segwit2x but not peered to the main group: Will not see valid blocks until the segwit2x chain overtakes the non-segwit2x one, but will not accept the invalid ones. UX degradation.
- Peered + signalling + orphaning miners: Business as usual.

As usual, knowing and enforcing the soft-fork rules is safer than being downgraded to SPV security, as long as a majority of hashpower is cooperating to enforce such rules.

Not clear to me why we'd want to discourage end users from running segwit2x nodes given the above. If anything we should be doing the opposite.

**Comment by istrau2** [comment]
@christophebiocca 
I don't think your math is correct. The following seems incorrect to me:
> This follows the same distribution as someone trying to do a 51% attach with minority hashpower

Aren't you forgetting about how connectivity factors into the equation here? I mean, if a non-BIP91 block is mined, all the BIP91 nodes ignore it and keep mining on top of the previous block. However, the non-BIP91 nodes do not ignore it and begin mining on top of the new block. Then, the non-BIP91 nodes are given what could be a significant head start, aren't they?

**Comment by istrau2** [comment]
In fact, this could possibly snowball (not sure how to calculate if a snowball effect is likely or not).

**Comment by christophebiocca** [comment]
They have a one block head start which is why I used the "odds of making 2 blocks" instead of the "odds of making 3 blocks" from the calculations. Because yes, the non-orphaning nodes have a 1-block head start each time. But that has nothing to do with connectivity, just has to do with the fact that it's the majority trying to orphan the minority, instead of the other way around.

This already accounts for the snowball effect. I encourage you to read section 11. of the whitepaper which explains the calculation in detail. I may have messed up the adaptation to this scenario. Specifically, the chance above is to get to a lead of 3 *before* being tied, but being tied does not immediately kill the chain, so the percentages might need to be scaled by `1 + q` to be truly accurate.

**Comment by istrau2** [comment]
@christophebiocca 

Yes, I do see what you are saying (regarding the headstart not being related to connectivity). Still, related to the headstart, it would seem that connectivity does still matter because it determines the extent of the headstart. I mean, given the issue at hand, it is only the hashrate of the largest connected network of BIP91 nodes that matters. In other words the numbers you mention are an ideal (if all BIP91 nodes are connected).

**Comment by christophebiocca** [comment]
I had assumed your choice of 20% instead of the more realistic 12% was an attempt to account for a partitioned network of segwit2x-enforcing miners leading to a lower hashrate.

20/80 is 12/48 which less than 4 biggest pools put together.
It's already a pretty big assumption that we wouldn't get any other pool to addnode one of the big 4.

**Comment by 0rmi** [comment]
@istrau2 Correct me if I'm wrong, but I believe that BIP91 miners are going to accept non-Segwit2x blocks as long as they signal "regular" segwit on bit1? So only miners with no signaling for segwit(of any kind) would have their blocks ignored, like BU miners for example. 

And such miners should be a miniscule fraction of the total hashpower and the likelihood of them mining any meaningful amount of blocks on top of each very small, thus not making this a particularly scary issue?

**Comment by ATBP** [comment]
Much more SegWit2x nodes should be connected to the network before miners even start signalling. Miners should be the last ones who upgrade to btc1 client, as correctly described inside SegWit2x Calendar at https://segwit2x.github.io/

July 14 - Agreement Participants Install and Test Milestone
July 21 - Nodes Running & Signaling begins

**Comment by istrau2** [comment]
@0rmi 

I don't actually know the answer to those questions. I am nowhere near an expert at this subject. Just trying to bring as much clarity as I can to the situation.

**Comment by istrau2** [comment]
@christophebiocca 

Ah, I see. Ok, makes sense. What do you think is the best solution to this issue?

**Comment by kek-coin** [comment]
@0rmi BIP91 enforces mandatory bit-1 signalling. This means that non-S2X miners will not be orphaned if they signal for OG-Segwit, while certain S2X miners will be orphaned if they don't start signalling bit-1 in time (ie. anyone with the 0x20000010 version; e.g. BTC.TOP, Antpool, BATPOOL, BTC.com, VIABTC are risking self-orphanage at this point).

**Comment by christophebiocca** [comment]
> S2X miners will be orphaned if they don't start signalling bit-1 in time (ie. anyone with the 0x20000010 version; e.g. BTC.TOP, Antpool, BATPOOL, BTC.com, VIABTC are risking self-orphanage at this point).

There's a lock-in period of ~3 days between the 80% threshold and the start of orphaning. They're not risking anything at this point unless they you think they'll all stop paying attention and go on a vacation for a few days right before the threshold is reached..

**Comment by kek-coin** [comment]
I didn't mention F2Pool did I?

**Comment by kek-coin** [comment]
On a more serious note, I'm not saying they will keep not signalling bit-1, I am just pointing out the mechanical properties of BIP91. I don't expect those pools to cause their own blocks to be orphaned like that, but just based on the current signal they would in fact be orphaned by BIP91 miners even though they signal for BIP91 activation.

**Comment by christophebiocca** [comment]
> What do you think is the best solution to this issue?

Going down the list of people who are most at risk of losing money:

- Non-bit1-signalling miners should prepare to signal. I suspect they're all quite ready to do so, but it'd be good to get confirmation from them.
- Non-orphaning miners should consider doing one of the following:
  - Run btc1.
  - Run @jameshilliard's segsignal repo.
  - Apply the necessary patches to their own code.
  - Blacklist the blocks that don't signal bit1 (either manually or with a script).
- Miners that aren't peered should connect to at least one miner, the bigger the better. FIBER is good too but it shouldn't become a single-point-of-failure.
- Merchant nodes not running segwit2x or bip91-nodes should either upgrade their nodes to enforce all consensus rules, or require more confirmations than usual.
- Users running segwit2x or bip91 don't need to do anything, just expect slower confirmations that usual.
- Peered + signalling + orphaning miners: Should help the non-peered ones get connected to them.

**Comment by istrau2** [comment]
@christophebiocca So you don't like @kek-coin 's proposal?

**Comment by jheathco** [comment]
While syncing activation of BIP91's orphaning of blocks to that of BIP148's would definitely help alleviate partitioning risk, @TheBlueMatt makes the point that it's pretty late in the game to be making these sorts of changes.  I'd tend to agree, although this would be an extremely small change to the codebase.

If @tomasvdw's description of the situation is accurate it seems that we may all be overstating this risk and it can be solved by simple peering changes.

**Comment by christophebiocca** [comment]
> So you don't like @kek-coin 's proposal?

I don't like any proposal that suggests solving a risk that can be fixed by a couple of `addnode` commands sometime this week, by instead introducing a risk that requires a full software upgrade to fix on an even shorter timeline.

It's measurably worse for everyone involved.

**Comment by kek-coin** [comment]
@christophebiocca I am not sure how a minimal change that will immensely improve the resilience of the system against malicious attacks while at the same time giving the best possible show of unificaiton by miners and users working together to avoid a chainsplit is measurably worse for everyone involved when the argument against it is that people are expected to upgrade - at a moment in time when no release even exists yet so people are expected to upgrade anyway.

**Comment by jgarzik** [comment]
It's possible for multiple people to be correct in this case :)

1) This is not a 5-alarm fire.  Big miners were already peered together before btc1 existed.  Miners remain well peered, are aware of the issue and doing even more to strengthen peering.  i.e. The equivalent of `addnode` is already occurring.

2) Peers with matching rulesets need to converge towards each other.  Helping that process along should be positive for btc1 and non-btc1 users alike.

**Comment by jheathco** [comment]
No PR here but change to sync BIP91 activation with 148 should be fairly straightforward in validation.cpp:
 
    int64_t nMedianTimePast = pindex->GetMedianTimePast();
    bool fBIP148Active = (nMedianTimePast >= 1501545600);

    // SEGWIT2X signalling.
    if (VersionBitsState(pindex->pprev, chainparams.GetConsensus(), Consensus::DEPLOYMENT_SEGWIT2X, versionbitscache) == THRESHOLD_ACTIVE &&
        VersionBitsState(pindex->pprev, chainparams.GetConsensus(), Consensus::DEPLOYMENT_SEGWIT,    versionbitscache) == THRESHOLD_STARTED &&
        fBIP148Active)
    {

Referencing https://github.com/bitcoin/bitcoin/compare/master...shaolinfry:bip-segwit-flagday - simply delays orphaning until at least 01 Aug 2017 00:00:00 UTC

**Comment by JaredR26** [comment]
> @JaredR26 to circle back on your psychological
> approach to bitcoin consensus rules; don't you
> think that miners and users joining forces to avoid a
> chainsplit might make for a mighty fine symbolic
> event?

I would heartily agree with this and gladly go for it if not for one
thing... A number of individuals have repeatedly and clearly attempted to
block or derail the 2mb hardfork to the point of approaching
irrationality.  This issue raised was one of the first that didn't reek of
an ulterior motive behind it to me.  (And I'm glad for that!)

If I felt like core could get on board and stop fighting just so they get
to draw the line in the sand, I'd be on board with what you're saying in a
heart beat.

As I said on the email list recently, to me this fight comes down to
keeping bitcoin whole instead of two "coins."  The big block community is
frothing at the mouth to fork off and block segwits activation.  If the
coming 2x hardfork doesn't have unity, there WILL be two bitcoins.  The big
blocker community will not defect back to the other fork for a minimum of 6
months, and much much longer(or forever) if they have enough miners and
users for a viable alternative.

Worse, from cores perspective, they might LOSE that bet even if they
maintain a viable chain- the big blocker coin might be the more popular and
cause defections from the legacy segwit coin.  I truly don't understand why
the fight against the consensus hardfork that could keep us together is so
strong, but so long as that goes on... I feel(and others, too) I have to
approach every proposal, every issue with suspicion.  It's not very
pleasant, but then again none of this scaling debate has been pleasant.

(FYI, I don't agree with your implicit assumption that segwit2x== miners
and UASF== users.  I think UASF is most in line with small blockers and
segwit2x is more in line with moderates. Disregarding that implication, I
do truly wish we could get everyone on board together as you stated)

**Comment by kek-coin** [comment]
BIP91 vs. BIP148 is explicitly MASF vs. UASF as the only difference is the activation method - Core-style hashrate signal vs. Satoshi-style flagday. I agree with you that there are users on the BIP91 side and miners on the BIP148 side, but my intention was more to point at the decentralized BIP148 p2p network helping out the hashrate powerhouses behind BIP91 connect to eachother and their users.

I don't think that your qualifier for UASF-supporters as "smallblockers" holds up to scrutiny, that's just a tired and beaten rbtc meme. If anything, UASFers are insisting that the miners finally activate that blocksize increase that's been waiting for their go-ahead for the better part of a year now. But this is a kind of discussion best left to other forums, let's stick to how to make things technically feasible here.

**Comment by christophebiocca** [comment]
> that will immensely improve the resilience of the system against malicious attacks

The evidence does not bear that out. Please explain what malicious attack you're thinking about and how throwing non-mining nodes at a mining consensus problem solves it.

> while at the same time giving the best possible show of unificaiton by miners and users working together to avoid a chainsplit

At the risk of potentially permanently chainsplitting users and miners if they don't follow this extremely closely and don't realise that the meaning of bit 4 has been changed *during the activation period*.

> when the argument against it is that people are expected to upgrade

Not that they're expected to upgrade but that they will *chain-split off the network permanently* if they don't, because we'd have changed the rules out from under them.

> at a moment in time when no release even exists yet so people are expected to upgrade anyway.

This is the [3rd](https://github.com/btc1/bitcoin/releases) release already, people are running it in the field.

Arguably the chances that a miner/user doesn't upgrade in time are low, but they're still higher than the chances of a miner not `addnode`'ing a peer in time, especially when that's something most of them have been doing for longer than they've been running segwit2x.

Also note that non-mining users running segwit2x are not at risk at all under the current activation sequence (they'll, at worst, have to wait for the "correct" chain to overtake the "broken" one, but they'll never accept the "broken" chain nor any payments on it).

**Comment by kek-coin** [comment]
> the meaning of bit 4 has been changed during the activation period

I'll point you to [the working group's calendar](https://segwit2x.github.io/) which clearly states

> July 21 - Nodes Running & **Signaling begins**

(Emphasis mine). The fact that nobody bothered to correct the starting date of June 1st in @jameshilliard's segsignal code does not change that.

> This is the 3rd release already, people are running it in the field.

Are you completely unable to differentiate between alphas, betas, RCs and releases? o.O

**Comment by JaredR26** [comment]
> I'll point you to the working group's calendar which clearly states

I mean, we've almost triggered in two activation periods in the last two days, there's threads all over reddit about it... But sure, schedule says we haven't started yet.

**Comment by kek-coin** [comment]
I had considered making a PR to fix the start date of June 1st and set it to July 21st as per the planning but I thought that miners would just follow the plan and it was a bit of a useless formality and get me chewed out for sidetracking the WG's attention :/

**Comment by JaredR26** [comment]
> and set it to July 21st as per the planning but I thought that miners would just follow the plan

I don't think it would have done anything.  Have you ever tried to herd cats?  The file release dropped, miners picked it up and began running it.  I doubt most of them were aware of the schedule or would be aware of any schedule changes if they were made.  :P

**Comment by kek-coin** [comment]
The release did not drop yet... But that only makes your cat-herding point stronger.

**Comment by christophebiocca** [comment]
> July 21 - Nodes Running & Signaling begins

Some people jumped the gun and are signalling ahead of schedule. I'm aware of that and have mentioned it in other contexts.

> Are you completely unable to differentiate between alphas, betas, RCs and releases? o.O

I was wondering whether you'd bring that up when I initially wrote it. I am aware that this is labelled RC. It is still a set of runnable binaries (not that miners have any problems with compiling code on their own) that have now been deployed in the wild.

So sure, we can take the prescriptive approach to software development, and pull the rug out from those silly miners for signalling ahead of schedule. Or we can instead *not do that*.

The first one is technically allowed. The second one is less disruptive (or less likely to be disruptive) to the people running the software and thus in my mind the better solution.

Now, there is new code you could release that won't harm miners that don't upgrade:
Recognize and enforce bit-4 activation, but don't signal it unless this is the last activation period ending before before August 1st (or any activation period thereafter).

This would keep signalling low for now, and only start miner orphaning shortly before the UASF nodes would start exclusively relaying such blocks. There's still the convergence risk but it'd be an issue only for a few days at most.

But frankly I wouldn't even do that much since even just being close to the UASF deadline presents its own risks, by allowing a few miners to derail the whole thing and cause a much nastier network split with very little warning, whether they do that to get miners to run UASF, UAHF, or just to harm their competitors. Much better to get this done nice and early, neuter the risk of any compatible-chain-split, and get ready for the next part of the roadmap.

**Comment by DannyDaemonic** [comment]
Has anyone tried contacting the UASF people to see if they'd be willing to add BIP91 support to their client? I realize that they are all about BIP148, but I think the bigger goal was activating segwit, so they may be willing to add BIP91 support.

**Comment by moneronoob12345** [comment]
@DannyDaemonic I think that is a bad idea as it will convey the idea that bit 4 is just for segwit, not segwit + 2MB HF

**Comment by TRUMP3030** [comment]
If bitcoin were to split..How does it affect the original maximum allowance of 21 million original BTC coins?
I have not been able to really find any information regarding that and you guys seem pretty much like geniuses. Sorry if it is off topic.

**Comment by benkloester** [comment]
@kek-coin, I now think I misunderstood you, but I originally took your suggestion to be to change the current UASF deployment so that orphaning starts not only on a flag day, but if BIP91 is active. So change the check in the UASF versions of Bitcoin to be if "BIP148 ACTIVE || BIP91 ACTIVE" {then orphan non-SW blocks}.

I think I took the "BIP148-nodes will help out to alleviate the partitioning danger" too literally perhaps, but I don't see any reason why doing it the way I originally understood wouldn't also work.

This would increase the number of nodes that would relay the blocks on the shorter-but-signalling-SW chain rather than the longer-not-signalling-SW chain, would it not? 

The changes to the UASF code would also be very simple, something like

    bool bip91Active = VersionBitsState(pindex->pprev, chainparams.GetConsensus(), Consensus::DEPLOYMENT_SEGWIT2X, versionbitscache) == THRESHOLD_ACTIVE &&
        VersionBitsState(pindex->pprev, chainparams.GetConsensus(), Consensus::DEPLOYMENT_SEGWIT,    versionbitscache) == THRESHOLD_STARTED;

    if ( ( (nMedianTimePast >= 1501545600) || bip91Active ) && // Tue 01 Aug 2017 00:00:00 UTC or BIP91 is active
    ... [etc]

in validation.cpp, plus the addition of SEGWIT2x as bit 4 in chainparams (if preferred the UASF code could even call it something else, as long as it looks for bit 4).
          
@moneronoob12345
The above solution carries no implications for the meaning of bit 4, in my opinion, since it doesn't involve the UASF nodes actually signalling bit 4 at all - they will just help "play sherriff" if and when BIP91 orphaning becomes active.

I think the symbolism is also good for the UASF, because they actually get to orphan non-SW signalling blocks this way, instead of being a "sherriff in a town with no criminals" come Aug 1st if BIP91 activates. AND they reduce partition / reorg risk at the same time.

**Comment by marthinus-engelbrecht** [comment]
@TRUMP3030 It won't affect it. The fork is not changing the cap. But there will be multiple coins with the same 21m cap. Maybe delete your comment now, then I'll delete mine.

**Comment by kek-coin** [comment]
@benkloester I meant the change in logic to be on the S2X/BIP91 side since it is a smaller group of nodes that would need to be upgraded - as can be seen on [saltylemon](https://uasf.saltylemon.org/) adoption rate of the 1.0 release with p2p improvements of the BIP148 code is not quick enough to help out here.

@DannyDaemonic there's already a PR for that by @luke-jr but it doesn't have the partitioning mitigation of the segsignal/bitcoin repository so I don't think it's a good idea to merge that yet.

**Comment by kek-coin** [comment]
But with how it seems that we might have BIP91 lock-in this signal period and @TheBlueMatt providing support on FIBRE I think that my suggestion is now moot.

**Comment by jgarzik** [comment]
Events, hashpower and changes such as #87 combine to address/render this issue moot.
